zkevm-circuits/src/compress_circuit.rs

```diff
@@ -0,0 +1,309 @@
+            });
+        }
+
+        // rpi from pi-circuit will be unfolded into 32 bytes
+        meta.create_gate("origin_rpi = rpi0 + rpi1 + ... + rpi31", |meta| {
+            let q_enable = meta.query_selector(q_enable);
+            let origin_rpi = meta.query_advice(args.raw_public_inputs, Rotation::cur());
+            let rpi = raw_public_inputs.map(|rpi| meta.query_advice(rpi, Rotation::cur()));
+            vec![q_enable * (origin_rpi - combine::expr(rpi))]
+        });
+
+        // cpi_rlc_acc
+        // TODO: use 32 bytes per rlc item
+        meta.create_gate(
+            "cpi_rlc_acc[0] = next_cpi_rlc_acc * rand_rpi + rlc(cpi0, ..., cpi31)",
+            |meta| {
+                let q_not_end = 1.expr() - meta.query_selector(q_end);
+                let cur_cpi_rlc_acc = meta.query_advice(cpi_rlc_acc, Rotation::cur());
+                let next_cpi_rlc_acc = meta.query_advice(cpi_rlc_acc, Rotation::next());
+                let rand_cpi = meta.query_advice(rand_cpi, Rotation::cur());
+
+                let rlc_cpi = rlc::expr(
+                    compressed_public_inputs.map(|cpi| meta.query_advice(cpi, Rotation::cur())),
+                    rand_cpi.clone(),
+                );
+
+                vec![q_not_end * (next_cpi_rlc_acc * rand_cpi + rlc_cpi - cur_cpi_rlc_acc)]
+            },
+        );
+
+        meta.create_gate("cpi_rlc_acc[last] = rlc(cpi0, ..., cpi31)", |meta| {
+            let q_end = meta.query_selector(q_end);
+            let rand_rpi = meta.query_advice(rand_cpi, Rotation::cur());
+            let rlc_cpi = rlc::expr(
+                compressed_public_inputs.map(|cpi| meta.query_advice(cpi, Rotation::cur())),
+                rand_rpi,
+            );
+            let cpi_rlc_acc = meta.query_advice(cpi_rlc_acc, Rotation::cur());
+            vec![q_end * (rlc_cpi - cpi_rlc_acc)]
+        });
+
+        // rand_rpi[i] == rand_rpi[j]
+        meta.create_gate("rand_pi = rand_rpi.next", |meta| {
+            // q_not_end * row.rand_rpi == q_not_end * row_next.rand_rpi
+            let q_not_end = 1.expr() - meta.query_selector(q_end);
+            let cur_rand_cpi = meta.query_advice(rand_cpi, Rotation::cur());
+            let next_rand_cpi = meta.query_advice(rand_cpi, Rotation::next());
+
+            vec![q_not_end * (cur_rand_cpi - next_rand_cpi)]
+        });
+
+        Self {
+            q_enable,
+            q_end,
+            raw_public_inputs,
+            compressed_public_inputs,
+            cpi_rlc_acc,
+            rand_cpi,
+            huffman_table,
+            _marker: PhantomData,
+        }
+    }
+}
+
+impl<F: Field> CompressCircuitConfig<F> {
+    fn load_huffman_table(&self, layouter: &mut impl Layouter<F>) -> Result<(), Error> {
+        layouter.assign_table(
+            || "fixed huffman table",
+            |mut table| {
+                for (value, code) in FIXED_HUFFMAN_TABLE.iter().enumerate() {
```

Comment:
To implement the scheme from https://github.com/taikoxyz/zkevm-circuits/issues/42#issuecomment-1351651634 you'll need 3 columns.

Reply:
So, with the new approach, we encode the data in circuit instead of decoding data. We only need lookup code by value without the length

Reply:
Ah yeah you could see see it as encoding the data. But still, isn't it similar in that you still need the length information to you know how to combine the bits into bytes? In the current table you don't know how many bits `code` has, so no way to know how many bits to jump ahead in the the encoded data stream?

Reply:
Yeah, you're right, now I just use the variable number of bits/row, and use rlc to encode them. But I found it is hard for contract to make a same public inputs from compressed txlist(can't get variable length bits).  It's a mistake.

So I will use the bytes stream for rlc encoding. 

Thank you very much!